users/src/main/java/io/crate/operation/user/Privileges.java

```diff
@@ -0,0 +1,143 @@
+import io.crate.analyze.user.Privilege;
+import io.crate.exceptions.AnalyzerUnknownException;
+import io.crate.exceptions.ColumnUnknownException;
+import io.crate.exceptions.PartitionUnknownException;
+import io.crate.exceptions.PermissionDeniedException;
+import io.crate.exceptions.RelationUnknownException;
+import io.crate.exceptions.RepositoryAlreadyExistsException;
+import io.crate.exceptions.RepositoryUnknownException;
+import io.crate.exceptions.SchemaUnknownException;
+import io.crate.exceptions.SnapshotAlreadyExistsExeption;
+import io.crate.exceptions.SnapshotUnknownException;
+import io.crate.exceptions.TableAliasSchemaException;
+import io.crate.exceptions.TableAlreadyExistsException;
+import io.crate.exceptions.TableUnknownException;
+import io.crate.exceptions.UserDefinedFunctionAlreadyExistsException;
+import io.crate.exceptions.UserDefinedFunctionUnknownException;
+
+class Privileges {
+
+    /**
+     * Checks if the user the concrete privilege for the given class and ident, if not raise exception.
+     */
+    static void raiseMissingPrivilegeException(Privilege.Type type,
+                                               Privilege.Clazz clazz,
+                                               String ident,
+                                               User user) throws PermissionDeniedException {
+        assert user != null : "User must not be null when trying to validate privileges";
+        assert type != null : "Privilege type must not be null";
+
+        if (user.hasPrivilege(type, clazz, ident) == false) {
+            throw new PermissionDeniedException(user.name(), type);
+        }
+    }
+
+    /**
+     * Checks if the user has ANY privilege for the given class and ident, if not raise exception.
+     */
+    @VisibleForTesting
+    static void raiseMissingPrivilegeException(Privilege.Clazz clazz,
+                                               String ident,
+                                               User user) throws PermissionDeniedException {
+        assert user != null : "User must not be null when trying to validate privileges";
+        if (user.hasAnyPrivilege(clazz, ident) == false) {
+            throw new PermissionDeniedException(user.name());
+        }
+    }
+
+    static void validateException(Throwable t, SessionContext sessionContext) {
+        User user = sessionContext.user();
+        if (user == null) {
+            // this can occur when the hba setting is not there,
+            // in this case there is no authentication and everyone
+            // can access the cluster
+            return;
+        }
+
+        if (user.isSuperUser()) {
+            return;
+        }
+        if (t instanceof TableUnknownException) {
+            String[] parts = ((TableUnknownException) t).getTableIdent().split("\\.");
+            String schemaName = parts.length > 1 ? parts[0] : sessionContext.defaultSchema();
+            raiseMissingPrivilegeException(Privilege.Clazz.SCHEMA, schemaName, user);
```

Comment:
We need `return;` statements after each `raiseMissingPrivilegeException` call (so we don't keep checking the other if conditions if we do not need to raise the permission exception. if we are to have some sort of exceptions hierarchy at some point this would actually be malicious). This is still error prone. 
Maybe convert to a `switch` statement - that way the "fall through" nature of the construct is visible by virtue of how switch works (and `break`ing is easier to track)?

I think this is messy because `raiseMissingPrivilegeException` doesn't actually always raise an exception.
Maybe it's worth thinking about renaming `raiseMissingPrivilegeException` to `getPermissionDeniedOrOriginalException` and make it *return* either a `PermissionDeniedException` or the passed in `t` exception. Then each `if`/`switch` body will say `throw getPermissionDeniedOrOriginalException(t, SCHEMA, schemaName, user);` (we throw the original exception anyway)

Or, less intrusive, have `raiseMissingPrivilegeException` renamed to `getPermissionDeniedException` and have it return Optional<PermissionDeniedException> - that will have the callers always check if they're getting one and throw it or break.